9.trees_and_graphs/task2.go

```diff
@@ -0,0 +1,582 @@
+// Реализуйте 2-3-дерево: вставка, поиск, удаление
+
+package main
+
+// Node interfaces
+type Finder interface {
+	Find(k int) bool
+}
+
+type Swapper interface {
+	Swap(x, y *int)
+}
+
+type Sorter interface {
+	Sort()
+	Sort2(x, y *int)
+	Sort3(x, y, z *int)
+}
+
+type Inserter interface {
+	InsertToNode(k int)
+}
+
+type Remover interface {
+	RemoveFromNode(k int)
+}
+
+type Transformer interface {
+	BecomeNode2(k int, first, second *Node)
+}
+
+type LeafChecker interface {
+	IsLeaf() bool
+}
+
+// Node struct
+type Node struct {
+	size   int
+	key    [3]int
+	first  *Node
+	second *Node
+	third  *Node
+	fourth *Node
+	parent *Node
+}
+
+// Implementing Node interface
+
+// Find method return true if key `k` on the vertex, else `false`
+func (n *Node) Find(k int) bool {
+	for i := 0; i < n.size; i++ {
+		if n.key[i] == k {
+			return true
+		}
+	}
+	return false
+}
+
+// Swap method
+func (n *Node) Swap(x, y *int) {
+	*x, *y = *y, *x
+}
+
+// Sort2 method
+func (n *Node) Sort2(x, y *int) {
+	if *x > *y {
+		n.Swap(x, y)
+	}
+}
+
+// Sort3 method
+func (n *Node) Sort3(x, y, z *int) {
+	if *x > *y {
+		n.Swap(x, y)
+	}
+	if *x > *z {
+		n.Swap(x, z)
+	}
+	if *y > *z {
+		n.Swap(y, z)
+	}
+}
+
+// Sort method
+func (n *Node) Sort() {
+	if n.size == 1 {
+		return
+	}
+	if n.size == 2 {
+		n.Sort2(&n.key[0], &n.key[1])
+	}
+	if n.size == 3 {
+		n.Sort3(&n.key[0], &n.key[1], &n.key[2])
+	}
+}
+
+// InsertToNode method insert key `k` to the vertex (not in tree)
+func (n *Node) InsertToNode(k int) {
+	n.key[n.size] = k
+	n.size++
+	n.Sort()
+}
+
+// RemoveFromNode method remove key `k` from the vertex (not from tree)
+func (n *Node) RemoveFromNode(k int) {
+	if n.size >= 1 && n.key[0] == k {
+		n.key[0] = n.key[1]
+		n.key[1] = n.key[2]
+		n.size--
+	} else if n.size == 2 && n.key[1] == k {
+		n.key[1] = n.key[2]
+		n.size--
+	}
+}
+
+// BecomeNode2 method transform to second vertex
+func (n *Node) BecomeNode2(k int, first, second *Node) {
+	n.key[0] = k
+	n.first = first
+	n.second = second
+	n.third = nil
+	n.fourth = nil
+	n.parent = nil
+	n.size = 1
+}
+
+// IsLeaf method whether the node is a leaf; validation is used on insertion and deletion
+func (n *Node) IsLeaf() bool {
+	return n.first == nil && n.second == nil && n.third == nil
+}
+
+// Split func devide a node if it contains 3 keys
+func Split(item *Node) *Node {
+	if item.size < 3 {
+		return item
+	}
+
+	x := &Node{
+		size:   1,
+		key:    [3]int{item.key[0]},
+		first:  item.first,
+		second: item.second,
+		parent: item.parent,
+	}
+	y := &Node{
+		size:   1,
+		key:    [3]int{item.key[2]},
+		first:  item.third,
+		second: item.fourth,
+		parent: item.parent,
+	}
+
+	switch {
+	case x.first != nil:
+		x.first.parent = x
+	case x.second != nil:
+		x.second.parent = x
+	}
+	switch {
+	case y.first != nil:
+		y.first.parent = y
+	case y.second != nil:
+		y.second.parent = y
+	}
```

Comment:
вот тут еще багуля: у тебя должны отработать все кейсы, если они существуют, а отрабатывает максимум 2 из 4 из-за свитча, либо надо было добавить continue, либо на ифах сделать
```suggestion
	if x.first != nil {
		x.first.parent = x
	}
	if x.second != nil {
		x.second.parent = x
	}
	if y.first != nil {
		y.first.parent = y
	}
	if y.second != nil {
		y.second.parent = y
	}

```